### ISSUE TITLE
Installation notes, pitfalls

### ISSUE THREAD

**Issue by drphilmarshall** [DarkEnergyScienceCollaboration/Twinkles#160]
Hi @jchiang87 - I am trying to reproduce your postage stamps, and documenting everything I need to do to be able to do that in [INSTALL.md](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/master/INSTALL.md). Can you please read through the steps I followed, and then comment on the problem I am hitting? 
```
ImportError: dlopen(/Users/pjm/miniconda2/opt/lsst/daf_persistence/python/lsst/daf/persistence/_persistenceLib.so, 10): Library not loaded: @rpath/libboost_system.dylib
  Referenced from: /Users/pjm/miniconda2/opt/lsst/boost/lib/libboost_filesystem.dylib
  Reason: image not found
```
I did all the setups as advised - have I got the wrong setup tags or something? Does it matter which order I do the `conda install` steps? Or do I just need an additional library beyond the Stack?

**Comment by drphilmarshall** [comment]
PS. That error came from the first cell of the [PostageStampMaker_demo.ipynb](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/master/examples/notebooks/PostageStampMaker_demo.ipynb)

**Comment by rbiswas4** [comment]
I think you have hit the kind of problems that has been bugging the whole galsim issue with conda. My impression was that this should not have happened within the stack, but maybe I am wrong. The problem is probably that you have the required dylb file (search for it) but your system is not looking for it in the right place (otool -L library) tells you where it is looking for it.

The quickest way out for now might be using install_name_tool as explained in 
http://thecourtsofchaos.com/2013/09/16/how-to-copy-and-relink-binaries-on-osx/

**Comment by drphilmarshall** [comment]
Well, it seems to be right there in the miniconda installation:
```
pjm@PPA-PC92478 > find miniconda2/. | grep libboost_filesystem.dylib
miniconda2/./opt/lsst/boost/lib/libboost_filesystem.dylib
miniconda2/./pkgs/lsst-boost-1.59.1-0/opt/lsst/boost/lib/libboost_filesystem.dylib
miniconda2/./pkgs/lsst-boost-1.59.5-3/opt/lsst/boost/lib/libboost_filesystem.dylib

pjm@PPA-PC92478 > otool -L miniconda2/./opt/lsst/boost/lib/libboost_filesystem.dylib
miniconda2/./opt/lsst/boost/lib/libboost_filesystem.dylib:
	@rpath/../opt/lsst/boost/lib/libboost_filesystem.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libboost_system.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

That `install_name_tool` cannot be the right thing to do, can it? I thought `conda` was the recommended way to go to get both the stack and `lsst_sims` - so this problem must have been encountered before. Is everyone doing this `install_name_tool` thing?

**Comment by rbiswas4** [comment]
Previously I thought the only place this happens is with galsim, but I am not sure the system is well tested enough. 
This maybe a problem I have led you down towards by suggesting the use of conda, which I like to use (eg. on lsstdev). Let us see if there are better ideas. @danielsf @cwwalter have spent a lot of time with the install_name_tool thing, but they might know if this happens at this stage.

**Comment by jchiang87** [comment]
@drphilmarshall I think the dylib you need to search for is `libboost_system.dylib` instead of `libboost_filesystem.dylib`, but I agree that having to use `install_name_tool` is not really an acceptable thing to have to do to get the stack working on your machine.  Also, from what I've been googling just now, it seems strange that the missing library is identified as `@rpath/libboost_system.dylib`, since the `@rpath` indicates it is already searching for that dylib in a number of places.  I would do the
```
find miniconda2/. | grep libboost_system.dylib
```
thing, and if it turns up, add the path to that file to your DYLD_LIBRARY_PATH env var as a workaround for now.

**Comment by drphilmarshall** [comment]
Thanks, you two! I fixed my DYLD_LIBRARY_PATH as suggested and that solved that problem. I then ran into some setup problems: should the DYLD_LIBRARY_PATH be fixed in the setup script? I thought not - so tried to put it in my .cshrc, and then ran into problems running the setup script from my .cshrc... 
Ive checked all this into the `conda` branch and opened a PR #164 so you can see my progress, such as it is. Anyway, the latest error is the following:
```
/Users/pjm/work/stronglensing/LSST/DESC/Twinkles/python/desc/twinkles/generatePhosimInput.py in <module>()
     20 from lsst.sims.catUtils.utils import ObservationMetaDataGenerator
     21 from lsst.sims.catalogs.generation.db import CatalogDBObject
---> 22 from lsst.sims.catalogs.generation.db.dbConnection import DBConnection
     23 from lsst.sims.catUtils.baseCatalogModels import OpSim3_61DBObject, StarObj, MsStarObj, \
     24         BhbStarObj, WdStarObj, RRLyStarObj, CepheidStarObj, GalaxyBulgeObj, GalaxyDiskObj, \

ImportError: cannot import name DBConnection
```
More Stack vs Sims conflict, perhaps?

**Comment by rbiswas4** [comment]
Did you do 
`setup sims_catUtils`?
or use tags to setup a tagged version if you have tagged them

**Comment by jchiang87** [comment]
Note that you can see everything that has been set up in your current configuration by doing
```
eups list
```

**Comment by danielsf** [comment]
If you are following the process documented here

https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/INSTALL.md

Then you have a problem:

The ../stable/ conda channel only contains v_11 of the DM stack and a very out-dated version of the sims stack. The sims stack conda installs live at

https://eupsforge.net/conda/dev

That would be why you are getting that failure at DBConnection (DBConnection is a relatively recent addition to the sims stack).

Try
```
cond config --add channels https://eupsforge.net/conda/dev
```
and then
```
conda install lsst_sims
```
you should get version 2.2.1, which will contain the DBConnection class

**Comment by drphilmarshall** [comment]
Thanks @danielsf - although so far it looks like I need to `conda install lsst-sims`, with a hyphen. The convention seems to be "hyphens for installing, underscores for setups" which I expect makes sense if you think about it long enough.

**Comment by drphilmarshall** [comment]
OK, looks like I need to switch channels as well as add a new one:
```
Fetching package metadata: ........SSL verification error: [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:590)
.SSL verification error: [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:590)
.
Solving package specifications: ...............................................
Warning: 2 possible package resolutions (only showing differing packages):
	[u'setuptools-git-1.1-py27_0.tar.bz2'],
	[u'setuptools-git-1.1-0.tar.bz2'],

# All requested packages already installed.
# packages in environment at /Users/pjm/miniconda2:
#
lsst-sims                 master.20150626170402.post157    ac1855abae_0    http://research.majuric.org/conda/stable/osx-64/lsst-sims-master.20150626170402.post157-ac1855abae_0.tar.bz2
```
How do I `conda install` from a particular channel in my list?

**Comment by danielsf** [comment]
(yeah.  Sorry.  I'm used to interacting with eups distrib install, which uses underscores)

**Comment by drphilmarshall** [comment]
!!

**Comment by danielsf** [comment]
Okay.  I was afraid of this.  Try
```
conda install lsst-sims sims-2.2.1
```
i.e. specify a version

**Comment by rbiswas4** [comment]
FYI: I did a conda install of sims on lsstdev to generate the instance catalogs for twinkles. As SNModel did not exist in older lsst_sims versions, I used:
```
source eups-setups.sh
setup sims_catUtils -t rbiswas -t b1887
source ${HOME}/.bash_profile
```
my `conda info` output is :
```
Current conda install:

             platform : linux-64
        conda version : 3.19.1
  conda-build version : 1.19.0
       python version : 2.7.11.final.0
     requests version : 2.9.1
     root environment : /home/rbiswas/src/conda-lsst/miniconda  (writable)
  default environment : /home/rbiswas/src/conda-lsst/miniconda
     envs directories : /home/rbiswas/src/conda-lsst/miniconda/envs
        package cache : /home/rbiswas/src/conda-lsst/miniconda/pkgs
         channel URLs : http://eupsforge.net/conda/dev/linux-64/
                        http://eupsforge.net/conda/dev/noarch/
                        https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /home/rbiswas/.condarc
    is foreign system : False
```

**Comment by danielsf** [comment]
@drphilmarshall 

Okay.... take a deep breath.... and try to resist the urge to throw something at me when you see how absurd this install command is....

```
conda install lsst-sims=master.20150928224422
```

(this is because lsst-sims isn't a meaningful package; it's just a script that ties all of the other sims packages together, so it doesn't usually get any helpful git tags attached to it; I will, of course, fix this when releasing future versions)

**Comment by danielsf** [comment]
incidentally the conda syntax for requesting a specific version of a package is
```
conda install packageName=packageVersion
```

**Comment by danielsf** [comment]
(I am going down this route because I am sure the current version of lsst-sims does not exist on the stable channel, so, by specifying a version, we should be able to resolve your channel ambiguity)

**Comment by drphilmarshall** [comment]
Thanks Scott - I'm at KIPAC tea with a damp towel on my forehead right now,
but I'll try that piece of wizardry in an hour or so. Thank you for helping
me! :-)

**Comment by drphilmarshall** [comment]
Alright! Got DB. But now I need SNObj:
```
/Users/pjm/work/stronglensing/LSST/DESC/Twinkles/python/desc/twinkles/generatePhosimInput.py in <module>()
     21 from lsst.sims.catalogs.generation.db import CatalogDBObject
     22 from lsst.sims.catalogs.generation.db.dbConnection import DBConnection
---> 23 from lsst.sims.catUtils.baseCatalogModels import OpSim3_61DBObject, StarObj, MsStarObj, \
     24         BhbStarObj, WdStarObj, RRLyStarObj, CepheidStarObj, GalaxyBulgeObj, GalaxyDiskObj, \
     25         GalaxyAgnObj, SNObj

ImportError: cannot import name SNObj
```
This was after I did the install you suggested, @danielsf, and the `setup sims_catUtils -t rbiswas -t b1887` you suggested, @rbiswas4 ! INSTALL.md is up to date with my working, in #164

**Comment by rbiswas4** [comment]
Aha. That is going to be harder ... not merged yet. The way I have worked is to mix this with a branch of sims_catUtils that has SNObj. We will try to push on the merge, but it will not be done by the time of the meeting. I have been working off a branch in sims_catUtils newTwinklesCatalog:

**Comment by jchiang87** [comment]
I've been dealing with the same problem.  Maybe as a temporary workaround we can put a try/except block in the `python/desc/twinkles/__init__.py` file, e.g., 
```
try:
    from generatePhosimInput import generatePhosimInput
except ImportError, eobj:
    print "Error importing generatePhosimInput", eobj
```

**Comment by drphilmarshall** [comment]
Happy to try that.

Hey Rahul, what is this about?
```
Ignoring unsupported tags in VRO: rbiswas

setup: Unsupported tag(s): rbiswas

```

**Comment by danielsf** [comment]
Would it be worthwhile to have an EUPS bootcamp sometime next week?

**Comment by jchiang87** [comment]
+100 I would definitely attend.

**Comment by jchiang87** [comment]
Maybe as a hack day thing?

**Comment by rbiswas4** [comment]
I am sorry .... there has been a miscommunication about the setup command, I should have added some more detail rather than sharing my config:


So here is what went into that:
When I did a conda install (and `source eups-setups.sh`) I had a particular version of lsst_sims using conda it had a tag b1887: 
I then cloned a copy of the sims_catUtils from github into a different location : for me this was:
`/home/rbiswas/src/sims_catUtils`. 
```
cd /home/rbiswas/src/sims_catUtils 
eups declare sims_catUtils -t tagName -r .
```
This declares a copy of the package sims_catUtils with a tag tagName in the location of current working directory (you can do it from elsewhere with an absolute path following -r) 
The usual recommendation for tagName is `$USER` which, for me, expands to rbiswas. I changed the branch on this repository to what I wanted 
```
git checkout newTwinklesCatalog
```
and then steup the sims_catUtils package with
```
setup sims_catUtils -t rbiswas -t b1887
```
Which sets up the package such that it uses the repo with tag rbiswas when it has available stuff, and B1887 when rbiswas does not. 
For you, you should have probably whatever your $USER expands to, and a tag of your installed package. I think this thing will disappear in that case.

**Comment by rbiswas4** [comment]
Maybe as part of DESchool?

**Comment by drphilmarshall** [comment]
Ah, I see - thanks Rahul. In the meantime, I have made the stamp making demo work, by try/except filtering of many sims modules. The question now is: should we merge #164? Come and see!

**Comment by rbiswas4** [comment]
Can I try to modify a couple of things on that?

**Comment by drphilmarshall** [comment]
Sure - can you do it by PR please, so I can see your changes? Thanks.

**Comment by jchiang87** [comment]
@drphilmarshall The DM team has a suggested fix for the `libboost_filesystem.dylib` loading problem:

https://community.lsst.org/t/library-loading-error-in-an-ipython-notebook/623

I'll make a note at #164 as well.

**Comment by drphilmarshall** [comment]
Revisiting this issue with @heather999 we decided to:
* Make a `doc/Setup` folder
* Merge together the conda [INSTALL.md](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/INSTALL.md) notes that I was working on, and the [Setup.md](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/conda/doc/Setup.md) that someone else wrote, into a simple .md file for "Getting_started_on_your_laptop.md"
* Write a separate .md file on "Getting_started_at_NERSC.md"

**Comment by heather999** [comment]
@drphilmarshall there is a new branch issue/160/installation where I added:
[Getting_started_on_your_laptop.rst](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/issue/160/installation/doc/Setup/Getting_started_on_your_laptop.rst)
It's an attempt at merging your Install.md and doc/Setup.md, as well as updating the doc for the current DMstack installation instructions.  I blindly took the instructions from Setup.md after the DMstack installation.  Currently, these instructions ignore the list of "Potential Pitfalls" listed in Install.md on the conda branch.  It would be good to see if these are still relevant and/or need to updated. It would also be great to test this out on your laptop!

**Comment by rbiswas4** [comment]
I was wondering if I could add a request if it falls into the natural play of things, otherwise we can try later. I was playing with this but did not quite finish:
 Can we split the travis script into an install script which the continuous integration script can call (and the user can call on his/her laptop as well) which will do this on a conda environment. The main advantage is 
- we could test on our laptop without out usual software getting in the way
- and the same script could be used by others to get started.

**Comment by drphilmarshall** [comment]
Sounds like a good idea, @rbiswas - what do you think, Heather?

Thanks for the notes, Heather! I trashed my miniconda2 folder and have
started rebuilding according to your instructions, embellishing them here
and there. Wish me luck! :-)

**Comment by drphilmarshall** [comment]
OK, I have some feedback for you, @heather999 - I'll open the pull request and put it there. You'll be able to see my edits that way, as well.

**Comment by drphilmarshall** [comment]
This issue was successfully addressed by #275 !
